thread-rcu/rcu.h

```diff
@@ -0,0 +1,280 @@
+        fprintf(stderr, "spin_unlock:pthread_mutex_unlock %d\n", ret);
+        abort();
+    }
+}
+
+#define current_tid() (unsigned int) pthread_self()
+
+/* Be careful here, since the C11 terms do no have the same sequential
+ * consistency  for the smp_mb(). Here we use the closely C11 terms,
+ * memory_order_seq_cst.
+ */
+#define smp_mb() atomic_thread_fence(memory_order_seq_cst)
+
+/* Compiler barrier, preventing the compiler from reordering memory accesses */
+#define barrier() __asm__ __volatile__("" : : : "memory")
+
+/* Don't use it directly without atomic operation.
+ * Otherwise, use READ_ONCE() or WRITE_ONCE().
+ */
+#define ACCESS_ONCE(x) (*(volatile _Atomic __typeof__(x) *) &(x))
+
+/* READ_ONCE() close to those of a C11 volatile memory_order_relaxed atomic read
+ * . However, for address, data, or control dependency chain, it is more like
+ * memory_order_consume. But, presently most of implement promote those kind of
+ * thing to memory_order_aquire.
+ */
+#define READ_ONCE(x)                                                     \
+    ({                                                                   \
+        barrier();                                                       \
+        __typeof__(x) ___x =                                             \
+            atomic_load_explicit(&ACCESS_ONCE(x), memory_order_consume); \
+        barrier();                                                       \
+        ___x;                                                            \
+    })
+
+/* WRITE_ONCE() quite close to C11 volatile memory_order_relaxed atomic store */
+#define WRITE_ONCE(x, val)                                                   \
+    do {                                                                     \
+        atomic_store_explicit(&ACCESS_ONCE(x), (val), memory_order_relaxed); \
+    } while (0)
+
+#include <errno.h>
+#include <stdatomic.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+
+#ifdef __CHECKER__
+#define __rcu __attribute__((noderef, address_space(__rcu)))
+#define rcu_check_sparse(p, space) ((void) (((typeof(*p) space *) p) == p))
+#define __force __attribute__((force))
+#define rcu_uncheck(p) ((__typeof__(*p) __force *) p)
+#define rcu_check(p) ((__typeof__(*p) __force __rcu *) p)
+#else
+#define __rcu
+#define rcu_check_sparse(p, space)
+#define __force
+#define rcu_uncheck(p) p
+#define rcu_check(p) p
+#endif /* __CHECKER__ */
+
+/* Avoid false sharing */
+#define CACHE_LINE_SIZE 64
+#define __rcu_aligned __attribute__((aligned(2 * CACHE_LINE_SIZE)))
+
+/* Per-thread variable
+ *
+ * rcu_nesting is to distinguish where the current thread is in which graced
+ * period. The reader will use the global variable rcu_nesting_idx to set up
+ * the corresponding index of rcu_nesting for the read-side critical section.
+ * It is a per-thread variable for optimizing the reader-side lock execution.
+ */
+struct rcu_node {
+    unsigned int tid;
+    int rcu_nesting[2];
```

Comment:
I suppose that `rcu_nesting` should be of boolean type, right? If so, we can even use bit-filed rather than integer array.